### ISSUE TITLE
OpenConfig implementation on IOS-XRv not supporting several options for BGP

### ISSUE THREAD

**Issue by jwbensley** [CiscoDevNet/openconfig-getting-started#3]
NB: This is the gRPC client "grpc_cfg.py" I am using based off of the public Cisco example: https://gist.github.com/jwbensley/cd2f5bfacee321328ca638b5b5523bd9

NB: This is using the IOS-XRv 6.2.1 stable vagrant image from https://devhub.cisco.com/artifactory/appdevci-release/XRv64/6.2.1/ dated 17th Oct 2016

Various OpenConfig types are note supported in the IOS-XRv 6.2.1 stable VM...

From openconfig-bgp-types.yang:

```
  typedef peer-type {
    type enumeration {
      enum INTERNAL {
        description "internal (iBGP) peer";
      }
      enum EXTERNAL {
        description "external (eBGP) peer";
      }
    }
    description
      "labels a peer or peer group as explicitly internal or
      external";
  }

  identity REMOVE_PRIVATE_AS_OPTION {
    description
      "Base identity for options for removing private autonomous
      system numbers from the AS_PATH attribute";
  }

  identity PRIVATE_AS_REMOVE_ALL {
    base REMOVE_PRIVATE_AS_OPTION;
    description
      "Strip all private autonmous system numbers from the AS_PATH.
      This action is performed regardless of the other content of the
      AS_PATH attribute, and for all instances of private AS numbers
      within that attribute.";
  }

  identity PRIVATE_AS_REPLACE_ALL {
    base REMOVE_PRIVATE_AS_OPTION;
    description
      "Replace all instances of private autonomous system numbers in
      the AS_PATH with the local BGP speaker's autonomous system
      number. This action is performed regardless of the other
      content of the AS_PATH attribute, and for all instances of
      private AS number within that attribute.";
  }

  typedef remove-private-as-option {
    type identityref {
      base REMOVE_PRIVATE_AS_OPTION;
    }
    description
      "set of options for configuring how private AS path numbers
      are removed from advertisements";
  }

```

I am using the code (generated with PyandBind v0.5.8):

```
from openconfig_bgp import openconfig_bgp
...
oc_bgp = openconfig_bgp()

oc_bgp.bgp.neighbors.neighbor[network["ipaddr4"]].config.enabled = "true"
oc_bgp.bgp.neighbors.neighbor[network["ipaddr4"]].config.remove_private_as = "PRIVATE_AS_REMOVE_ALL"
oc_bgp.bgp.neighbors.neighbor[network["ipaddr4"]].config.peer_type = "EXTERNAL"
```

From the below config none the following are supported:

```
"enabled": true, 
"remove-private-as": "PRIVATE_AS_REMOVE_ALL", 
"peer-type": "EXTERNAL"
```

Example config that works once these three above options are removed:

```
bensley@ubuntu-laptop:~/Python/pyb_oc/ios-xr-grpc-python-master/examples$ cat snips/broken_private_as.json 
{
    "openconfig-bgp:bgp": {
        "neighbors": {
            "neighbor": [
                {
                    "neighbor-address": "193.37.144.1", 
                    "config": {
                        "neighbor-address": "193.37.144.1", 
                        "description": "i3b gmbh", 
                        "peer-as": 39912,
                        "enabled": true,
                        "peer-group": "public_v4_peers", 
                        "remove-private-as": "PRIVATE_AS_REMOVE_ALL",
                        "peer-type": "EXTERNAL"
                    }, 
                    "afi-safis": {
                        "afi-safi": [
                            {
                                "afi-safi-name": "openconfig-bgp-types:ipv4-unicast", 
                                "config": {
                                    "enabled": true, 
                                    "afi-safi-name": "openconfig-bgp-types:ipv4-unicast"
                                }, 
                                "ipv4-unicast": {
                                    "prefix-limit": {
                                        "config": {
                                            "shutdown-threshold-pct": 90, 
                                            "restart-timer": 1, 
                                            "max-prefixes": 200
                                        }
                                    }
                                }
                            }
                        ]
                    }
                }
            ]
        }
    }
}
```

That config returns the error:

`{u'cisco-grpc:errors': {u'error': {u'error-type': u'protocol', u'error-severity': u'error', u'error-tag': u'unknown-element', u'error-message': u'Unknown element is specified.', u'error-info': {u'bad-element': u'enabled'}}}}`

After removing `"enabled": true` I get:

`{u'cisco-grpc:errors': {u'error': {u'error-type': u'protocol', u'error-severity': u'error', u'error-tag': u'unknown-element', u'error-message': u'Unknown element is specified.', u'error-info': {u'bad-element': u'peer-type'}}}}`

After then removing `"peer-type": "EXTERNAL"` something really strange happens, with the only unsupported option left being the `"remove-private-as": "PRIVATE_AS_REMOVE_ALL"` option, for some reason the config is accepted however the remove-private-as option isn't configured and instead very strangely all of the possible BGP AFI/SAFIs are configured towards this neighbour?!?

You can also see here that all those additional address families have been enabled globally to support this new neighbor which could have a very bad effect in a live production environment:

```
bensley@ubuntu-laptop:~/Python/pyb_oc/ios-xr-grpc-python-master/examples$ cat snips/broken_private_as.json 
{
    "openconfig-bgp:bgp": {
        "neighbors": {
            "neighbor": [
                {
                    "neighbor-address": "193.37.144.1", 
                    "config": {
                        "neighbor-address": "193.37.144.1", 
                        "description": "i3b gmbh", 
                        "peer-as": 39912, 
                        "peer-group": "public_v4_peers", 
                        "remove-private-as": "PRIVATE_AS_REMOVE_ALL"
                    }, 
                    "afi-safis": {
                        "afi-safi": [
                            {
                                "afi-safi-name": "openconfig-bgp-types:ipv4-unicast", 
                                "config": {
                                    "enabled": true, 
                                    "afi-safi-name": "openconfig-bgp-types:ipv4-unicast"
                                }, 
                                "ipv4-unicast": {
                                    "prefix-limit": {
                                        "config": {
                                            "shutdown-threshold-pct": 90, 
                                            "restart-timer": 1, 
                                            "max-prefixes": 200
                                        }
                                    }
                                }
                            }
                        ]
                    }
                }
            ]
        }
    }
}

bensley@ubuntu-laptop:~/Python/pyb_oc/ios-xr-grpc-python-master/examples$ python grpc_cfg.py 
Config before merge:

{
    "Cisco-IOS-XR-ipv4-bgp-cfg:bgp": {
        "instance": [
            {
                "instance-as": [
                    {
                        "four-byte-as": [
                            {
                                "as": 65001, 
                                "bgp-running": [
                                    null
                                ], 
                                "default-vrf": {
                                    "global": {
                                        "global-afs": {
                                            "global-af": [
                                                {
                                                    "af-name": "ipv4-unicast", 
                                                    "enable": [
                                                        null
                                                    ]
                                                }, 
                                                {
                                                    "af-name": "ipv6-unicast", 
                                                    "enable": [
                                                        null
                                                    ]
                                                }
                                            ]
                                        }, 
                                        "router-id": "1.1.0.10"
                                    }, 
                                    "bgp-entity": {
                                        "neighbor-groups": {
                                            "neighbor-group": [
                                                {
                                                    "neighbor-group-afs": {
                                                        "neighbor-group-af": [
                                                            {
                                                                "af-name": "ipv4-unicast", 
                                                                "activate": [
                                                                    null
                                                                ], 
                                                                "route-policy-in": "public_v4_peering_ingress", 
                                                                "route-policy-out": "public_v4_peering_egress"
                                                            }
                                                        ]
                                                    }, 
                                                    "create": [
                                                        null
                                                    ], 
                                                    "neighbor-group-name": "public_v4_peers"
                                                }, 
                                                {
                                                    "neighbor-group-afs": {
                                                        "neighbor-group-af": [
                                                            {
                                                                "af-name": "ipv6-unicast", 
                                                                "activate": [
                                                                    null
                                                                ], 
                                                                "route-policy-in": "public_v6_peering_ingress", 
                                                                "route-policy-out": "public_v6_peering_egress"
                                                            }
                                                        ]
                                                    }, 
                                                    "create": [
                                                        null
                                                    ], 
                                                    "neighbor-group-name": "public_v6_peers"
                                                }
                                            ]
                                        }
                                    }
                                }
                            }
                        ], 
                        "as": 0
                    }
                ], 
                "instance-name": "default"
            }
        ]
    }
}
Config after merge:

{
    "Cisco-IOS-XR-ipv4-bgp-cfg:bgp": {
        "instance": [
            {
                "instance-as": [
                    {
                        "four-byte-as": [
                            {
                                "as": 65001, 
                                "bgp-running": [
                                    null
                                ], 
                                "default-vrf": {
                                    "global": {
                                        "global-afs": {
                                            "global-af": [
                                                {
                                                    "af-name": "ipv4-unicast", 
                                                    "enable": [
                                                        null
                                                    ]
                                                }, 
                                                {
                                                    "af-name": "ipv4-multicast", 
                                                    "enable": [
                                                        null
                                                    ]
                                                }, 
                                                {
                                                    "af-name": "vp-nv4-unicast", 
                                                    "enable": [
                                                        null
                                                    ]
                                                }, 
                                                {
                                                    "af-name": "ipv6-unicast", 
                                                    "enable": [
                                                        null
                                                    ]
                                                }, 
                                                {
                                                    "af-name": "vp-nv6-unicast", 
                                                    "enable": [
                                                        null
                                                    ]
                                                }, 
                                                {
                                                    "af-name": "l2vpnvpls", 
                                                    "enable": [
                                                        null
                                                    ]
                                                }, 
                                                {
                                                    "af-name": "l2vpnevpn", 
                                                    "enable": [
                                                        null
                                                    ]
                                                }, 
                                                {
                                                    "af-name": "vp-nv4-multicast", 
                                                    "enable": [
                                                        null
                                                    ]
                                                }, 
                                                {
                                                    "af-name": "vp-nv6-multicast", 
                                                    "enable": [
                                                        null
                                                    ]
                                                }
                                            ]
                                        }, 
                                        "router-id": "1.1.0.10"
                                    }, 
                                    "bgp-entity": {
                                        "neighbor-groups": {
                                            "neighbor-group": [
                                                {
                                                    "neighbor-group-afs": {
                                                        "neighbor-group-af": [
                                                            {
                                                                "af-name": "ipv4-unicast", 
                                                                "activate": [
                                                                    null
                                                                ], 
                                                                "route-policy-in": "public_v4_peering_ingress", 
                                                                "route-policy-out": "public_v4_peering_egress"
                                                            }
                                                        ]
                                                    }, 
                                                    "create": [
                                                        null
                                                    ], 
                                                    "neighbor-group-name": "public_v4_peers"
                                                }, 
                                                {
                                                    "neighbor-group-afs": {
                                                        "neighbor-group-af": [
                                                            {
                                                                "af-name": "ipv6-unicast", 
                                                                "activate": [
                                                                    null
                                                                ], 
                                                                "route-policy-in": "public_v6_peering_ingress", 
                                                                "route-policy-out": "public_v6_peering_egress"
                                                            }
                                                        ]
                                                    }, 
                                                    "create": [
                                                        null
                                                    ], 
                                                    "neighbor-group-name": "public_v6_peers"
                                                }
                                            ]
                                        }, 
                                        "neighbors": {
                                            "neighbor": [
                                                {
                                                    "neighbor-address": "193.37.144.1", 
                                                    "remote-as": {
                                                        "as-xx": 0, 
                                                        "as-yy": 39912
                                                    }, 
                                                    "neighbor-group-add-member": "public_v4_peers", 
                                                    "description": "i3b gmbh", 
                                                    "neighbor-afs": {
                                                        "neighbor-af": [
                                                            {
                                                                "af-name": "ipv4-unicast", 
                                                                "activate": [
                                                                    null
                                                                ], 
                                                                "maximum-prefixes": {
                                                                    "discard-extra-paths": false, 
                                                                    "warning-percentage": 90, 
                                                                    "warning-only": false, 
                                                                    "prefix-limit": 200, 
                                                                    "restart-time": 1
                                                                }
                                                            }, 
                                                            {
                                                                "af-name": "ipv4-multicast", 
                                                                "activate": [
                                                                    null
                                                                ]
                                                            }, 
                                                            {
                                                                "af-name": "vp-nv4-unicast", 
                                                                "activate": [
                                                                    null
                                                                ]
                                                            }, 
                                                            {
                                                                "af-name": "ipv6-unicast", 
                                                                "activate": [
                                                                    null
                                                                ]
                                                            }, 
                                                            {
                                                                "af-name": "vp-nv6-unicast", 
                                                                "activate": [
                                                                    null
                                                                ]
                                                            }, 
                                                            {
                                                                "af-name": "l2vpnvpls", 
                                                                "activate": [
                                                                    null
                                                                ]
                                                            }, 
                                                            {
                                                                "af-name": "l2vpnevpn", 
                                                                "activate": [
                                                                    null
                                                                ]
                                                            }, 
                                                            {
                                                                "af-name": "vp-nv4-multicast", 
                                                                "activate": [
                                                                    null
                                                                ]
                                                            }, 
                                                            {
                                                                "af-name": "vp-nv6-multicast", 
                                                                "activate": [
                                                                    null
                                                                ]
                                                            }
                                                        ]
                                                    }
                                                }
                                            ]
                                        }
                                    }
                                }
                            }
                        ], 
                        "as": 0
                    }
                ], 
                "instance-name": "default"
            }
        ]
    }
}
```
Renaming that option to be called "remove-private-as-option" as per this OpenConfig typdef:
```

  typedef remove-private-as-option {
    type identityref {
      base REMOVE_PRIVATE_AS_OPTION;
    }
    description
      "set of options for configuring how private AS path numbers
      are removed from advertisements";
  }
```

just returns the error:

`{u'cisco-grpc:errors': {u'error': {u'error-type': u'protocol', u'error-severity': u'error', u'error-tag': u'unknown-element', u'error-message': u'Unknown element is specified.', u'error-info': {u'bad-element': u'remove-private-as-option'}}}}
`

**Comment by jwbensley** [comment]
If I modify the Example.get method in the grpc_cfg.py script (https://gist.github.com/jwbensley/cd2f5bfacee321328ca638b5b5523bd9) to pull all BGP config not using the Cisco specific YANG models;

`path = '{"bgp:bgp": [null]}'`

And also config a BGP peer manually with the remove-private-as option configured:

```
RP/0/RP0/CPU0:ios#show run router bgp
Wed Nov  9 17:40:04.743 UTC
router bgp 65001
 bgp router-id 1.1.0.10
 address-family ipv4 unicast
 !
 address-family ipv6 unicast
 !
 neighbor-group public_v4_peers
  address-family ipv4 unicast
   route-policy public_v4_peering_ingress in
   route-policy public_v4_peering_egress out
  !
 !
 neighbor-group public_v6_peers
  address-family ipv6 unicast
   route-policy public_v6_peering_ingress in
   route-policy public_v6_peering_egress out
  !
 !
!
RP/0/RP0/CPU0:ios#
RP/0/RP0/CPU0:ios#conf t
Wed Nov  9 17:40:06.806 UTC
RP/0/RP0/CPU0:ios(config)#router bgp 65001
RP/0/RP0/CPU0:ios(config-bgp)# neighbor 193.37.144.1
RP/0/RP0/CPU0:ios(config-bgp-nbr)#  remote-as 39912
RP/0/RP0/CPU0:ios(config-bgp-nbr)#  use neighbor-group public_v4_peers
RP/0/RP0/CPU0:ios(config-bgp-nbr)#  description i3b gmbh
RP/0/RP0/CPU0:ios(config-bgp-nbr)#  address-family ipv4 unicast
RP/0/RP0/CPU0:ios(config-bgp-nbr-af)#remove-private-AS 
RP/0/RP0/CPU0:ios(config-bgp-nbr-af)#commit
Wed Nov  9 17:40:32.956 UTC
RP/0/RP0/CPU0:Nov  9 17:40:33.282 : config[67585]: %MGBL-CONFIG-6-DB_COMMIT : Configuration committed by user 'vagrant'. Use 'show configuration commit changes 1000000049' to view the changes. 
RP/0/RP0/CPU0:ios#show run router bgp
Wed Nov  9 17:40:35.122 UTC
router bgp 65001
 bgp router-id 1.1.0.10
 address-family ipv4 unicast
 !
 address-family ipv6 unicast
 !
 neighbor-group public_v4_peers
  address-family ipv4 unicast
   route-policy public_v4_peering_ingress in
   route-policy public_v4_peering_egress out
  !
 !
 neighbor-group public_v6_peers
  address-family ipv6 unicast
   route-policy public_v6_peering_ingress in
   route-policy public_v6_peering_egress out
  !
 !
 neighbor 193.37.144.1
  remote-as 39912
  use neighbor-group public_v4_peers
  description i3b gmbh
  address-family ipv4 unicast
   remove-private-AS
  !
 !
!
```

We can now see when pulling the running BGP config over gRPC that the remove-private-as option is not even shown there, I guess it's not supported and/or implemented yet?

```
bensley@ubuntu-laptop:~/Python/pyb_oc/ios-xr-grpc-python-master/examples$ python grpc_cfg.py 
Config before merge:

{
    "openconfig-bgp:bgp": {
        "neighbors": {
            "neighbor": [
                {
                    "neighbor-address": "193.37.144.1", 
                    "config": {
                        "neighbor-address": "193.37.144.1", 
                        "peer-as": 39912, 
                        "peer-group": "public_v4_peers", 
                        "description": "i3b gmbh"
                    }, 
                    "afi-safis": {
                        "afi-safi": [
                            {
                                "config": {
                                    "enabled": true, 
                                    "afi-safi-name": "openconfig-bgp-types:ipv4-unicast"
                                }, 
                                "afi-safi-name": "openconfig-bgp-types:ipv4-unicast"
                            }
                        ]
                    }
                }
            ]
        }, 
        "peer-groups": {
            "peer-group": [
                {
                    "config": {
                        "peer-group-name": "public_v4_peers"
                    }, 
                    "peer-group-name": "public_v4_peers", 
                    "afi-safis": {
                        "afi-safi": [
                            {
                                "apply-policy": {
                                    "config": {
                                        "import-policy": [
                                            "public_v4_peering_ingress"
                                        ], 
                                        "export-policy": [
                                            "public_v4_peering_egress"
                                        ]
                                    }
                                }, 
                                "config": {
                                    "enabled": true, 
                                    "afi-safi-name": "openconfig-bgp-types:ipv4-unicast"
                                }, 
                                "afi-safi-name": "openconfig-bgp-types:ipv4-unicast"
                            }
                        ]
                    }
                }, 
                {
                    "config": {
                        "peer-group-name": "public_v6_peers"
                    }, 
                    "peer-group-name": "public_v6_peers", 
                    "afi-safis": {
                        "afi-safi": [
                            {
                                "apply-policy": {
                                    "config": {
                                        "import-policy": [
                                            "public_v6_peering_ingress"
                                        ], 
                                        "export-policy": [
                                            "public_v6_peering_egress"
                                        ]
                                    }
                                }, 
                                "config": {
                                    "enabled": true, 
                                    "afi-safi-name": "openconfig-bgp-types:ipv6-unicast"
                                }, 
                                "afi-safi-name": "openconfig-bgp-types:ipv6-unicast"
                            }
                        ]
                    }
                }
            ]
        }, 
        "global": {
            "config": {
                "as": 65001, 
                "router-id": "1.1.0.10"
            }, 
            "afi-safis": {
                "afi-safi": [
                    {
                        "config": {
                            "enabled": true, 
                            "afi-safi-name": "openconfig-bgp-types:ipv4-unicast"
                        }, 
                        "afi-safi-name": "openconfig-bgp-types:ipv4-unicast"
                    }, 
                    {
                        "config": {
                            "enabled": true, 
                            "afi-safi-name": "openconfig-bgp-types:ipv6-unicast"
                        }, 
                        "afi-safi-name": "openconfig-bgp-types:ipv6-unicast"
                    }
                ]
            }
        }
    }
}
```

**Comment by 111pontes** [comment]
@jwbensley Make sure you match the model version supported by the device (https://github.com/openconfig/public/tree/bfb2695faed5a2bc0dd2ce14b2fc15f6abc713ae/release/models/bgp).  We're currently enhancing gRPC, so you can check the model version programmatically in the same way you can do it with NETCONF.  Let me look at the AFI/SAFI issue.  It seems a new defect needs to be filed.  Will confirm

**Comment by 111pontes** [comment]
Had a chance to look at this in more detail.  Here are my findings on the three items you brough up:
1. `"enabled": true`
This is a leaf not defined in the version of the model currently available in XR.  We're in the process of upgrading the implementation.  It's important that you match model versions or watch for backward compatible leaves.
2. `"peer-type": "EXTERNAL"`
That leave is already declared in the deviation as not-supported.
3. `"remove-private-as": "PRIVATE_AS_REMOVE_ALL"` 
XR BGP has more granular behavior for this configuration and that creates a mapping challange.  I've filed a defect (CSCvc11493) for a better mapping to be considered.
Hope this clarifies.

**Comment by jwbensley** [comment]
@111pontes Again, thanks very much for your help here :)

 For 2. I will see if there is a way I can ingest the Cisco deviation YANG modules and ways of automatically detecting that.

For 1. & 3. I will wait and see what happens in future builds.

**Comment by 111pontes** [comment]
CSCvc11493 has been fixed.  Fix is expected in XR 6.2.1.  Points 1 and 2 reflect correct behavior given current implementation.
